File: src/git_forge_release_migrator/providers/registry.py
from __future__ import annotations

from dataclasses import dataclass

from .base import ProviderAdapter
from .bitbucket import BitbucketAdapter
from .github import GitHubAdapter
from .gitlab import GitLabAdapter


@dataclass
class ProviderRegistry:
    adapters: dict[str, ProviderAdapter]

    @classmethod
    def default(cls) -> "ProviderRegistry":
        return cls(
            adapters={
                "github": GitHubAdapter(),
                "gitlab": GitLabAdapter(),
                "bitbucket": BitbucketAdapter(),
            }
        )

    def get(self, provider: str) -> ProviderAdapter:
        if provider not in self.adapters:
            raise ValueError(f"Unsupported provider: {provider}")
        return self.adapters[provider]

    def pair_status(self, source: str, target: str) -> str:
        known = {"github", "gitlab", "bitbucket"}
        if source not in known or target not in known:
            return "unsupported"
        if source == target:
            return "unsupported"
        if source in known and target in known:
            return "enabled"
        return "unsupported"

    def require_supported_pair(self, source: str, target: str) -> None:
        status = self.pair_status(source, target)
        if status == "enabled":
            return
        raise ValueError(f"Provider pair {source}->{target} is unsupported.")

File: dart_cli/lib/src/providers/registry.dart
import '../core/adapters/provider_adapter.dart';
import 'bitbucket.dart';
import 'github.dart';
import 'gitlab.dart';

class ProviderRegistry {
  ProviderRegistry(this.adapters);

  final Map<String, ProviderAdapter> adapters;

  factory ProviderRegistry.defaults() {
    return ProviderRegistry(<String, ProviderAdapter>{
      'github': GitHubAdapter(),
      'gitlab': GitLabAdapter(),
      'bitbucket': BitbucketAdapter(),
    });
  }

  ProviderAdapter get(String provider) {
    final ProviderAdapter? adapter = adapters[provider];
    if (adapter == null) {
      throw ArgumentError('Unsupported provider: $provider');
    }

    return adapter;
  }

  String pairStatus(String source, String target) {
    const Set<String> known = <String>{'github', 'gitlab', 'bitbucket'};
    if (!known.contains(source) || !known.contains(target)) {
      return 'unsupported';
    }

    if (source == target) {
      return 'unsupported';
    }

    return 'enabled';
  }

  void requireSupportedPair(String source, String target) {
    if (pairStatus(source, target) != 'enabled') {
      throw ArgumentError('Provider pair $source->$target is unsupported.');
    }
  }
}
